README
File::Same

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

DESCRIPTION

File::Same allows  you  to compare a whole bunch of files (or directories full of
files) for "sameness".

Essentially this  was written to detect which files in a multitude of directories
are the same, and allow us to do further processing on those files.

INSTALLATION

To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

Replace make for nmake on Win32.

DEPENDENCIES

This module requires these other modules and libraries:

    Digest::MD5
    File::Spec (included with core perl)

COPYRIGHT AND LICENCE

Copyright (C) 2001 MessageLabs Limited.

Copyright (C) 2015 Mohammad S Anwar.

This is free software, you  may  use it and distribute it under the same terms as
Perl itself.

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Perl extension File::Same.

0.10   2016-03-27 MANWAR
       - Added standard test scripts t/manifest.t and t/pod.t.

0.09   2016-03-15 MANWAR
       - Added key 'provides' to the Makefile.PL script.

0.08   2015-12-25 MANWAR
       - Added section REPOSITORY to the pod document.

0.07   2015-12-24 MANWAR
       - Updated Makefile.PL script w.r.t LICENSE information.
       - Updated LICENSE file w.r.t. Copyright information.

0.06   2015-12-24 MANWAR
       - Updated Copyright and LICENSE information in the pod file.
       - Tidied up README file.

0.05   2015-12-23 MANWAR
       - Added LICENSE file.
       - Added MANIFEST.SKIP file.
       - Added LICENSE/BUGS/SUPPORT sections to the pod document.

0.04   2015-12-22
       - [NEILB]  Moved Same.pm to lib/File/Same.pm
       - [NEILB]  In SEE ALSO, make Digest::MD5 a link, and added File::Find::Duplicates
       - [NEILB]  Added "use warnings", and min perl version to code
       - [NEILB]  Added prereqs to Makefile.PL
       - [NEILB]  Changed dates in this file to be ISO 8601 format, as suggested by CPAN::Changes::Spec
       - [MANWAR] Bumped version in the package File::Same.
       - [MANWAR] Added key 'LICENSE' and 'MIN_PERL_VERSION' to Makefile.PL script.

0.03   2015-12-21 MANWAR
       - Adopted the module with kind permission from Matt Sergeant.

0.02   2001-11-06
       - binmode() for Win32
       - Forgot to close filehandles (oops!)

0.01   2001-10-29
       - original version; created by h2xs 1.21 with options -A -X -n File::Same.

lib/File/Same.pm
package File::Same;

$File::Same::VERSION   = '0.10';
$File::Same::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

File::Same - Detect which files are the same as a given one.

=head1 VERSION

Version 0.10

=cut

use 5.006;
use strict; use warnings;
use Digest::MD5;
use File::Spec;

my %md5s;

=head1 DESCRIPTION

File::Same uses MD5 sums to decide which files are the same in a given directory,
set of directories or set of files. It was originally written to test which files
are the same picture in multiple directories or under multiple filenames, but can
be generally useful for other systems.

File::Same will use an internal cache, for performance reasons.

File::Same will also be careful not to return C<$original> in the list of matched
files.

All of the functions return a list of files that match, with full path expanded.

=head1 SYNOPSIS

    use strict; use warnings;
    use File::Same;

    my @same1 = File::Same::scan_dirs('sample.txt', ['other', '.']);

    my @same2 = File::Same::scan_files('sample.txt', ['ex1.txt', 'ex2.txt']);

    my @same3 = File::Same::scan_dir('sample.txt', 'somedir');

=head1 METHODS

=head2 scan_files($original, \@list)

Scan a list of files to compare against a given file.

=cut

sub scan_files {
    my ($original, $files) = @_;

    my @results;
    my $orig_md5 = $md5s{$original};

    if (!$orig_md5) {
        my $ctx = Digest::MD5->new();
        open(FILE, $original) || die "Cannot open '$original' : $!";
        binmode(FILE);
        $ctx->addfile(*FILE);
        $orig_md5 = $ctx->hexdigest;
        close(FILE);
    }

    foreach my $file (@$files) {
        if (my $md5 = $md5s{$file}) {
            if ($orig_md5 eq $md5) {
                push @results, $file;
            }
        }
        else {
            my $ctx = Digest::MD5->new();
            open(FILE, $file) || die "Cannot open '$file' : $!";
            binmode(FILE);
            $ctx->addfile(*FILE);
            if ($orig_md5 eq $ctx->hexdigest) {
                push @results, $file;
            }
            close(FILE);
        }
    }

    return grep {_not_same($_, $original)} @results;
}

=head2 scan_dir($original, $dir)

Scan an entire directory to find files the same as this one.

=cut

sub scan_dir {
    my ($original, $dir) = @_;

    opendir(DIR, $dir) || die "Cannot opendir '$dir' : $!";
    my @files = grep { -f } map { File::Spec->catfile($dir, $_) } readdir(DIR);
    closedir(DIR);

    return scan_files($original, \@files);
}

=head2 scan_dirs($original, \@dirs)

Scan a list of directories to find files the same as this one.

=cut

sub scan_dirs {
    my ($original, $dirs) = @_;

    my @results;

    foreach my $dir (@$dirs) {
        push @results, scan_dir($original, $dir);
    }

    return @results;
}

#
#
# PRIVATE METHODS

sub _not_same {
    my ($file, $orig) = @_;

    return 0
        if (File::Spec->rel2abs($file) eq File::Spec->rel2abs($orig));

    return 1;
}

=head1 AUTHOR

=over 4

=item Original author Matt Sergeant, C<< <matt at sergeant.org> >>

=item Currently maintained by Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=back

=head1 REPOSITORY

L<https://github.com/manwar/File-Same>

=head1 SEE ALSO

L<Digest::MD5> - used to generate a checksum for every file.

L<File::Find::Duplicates> - another that can be used to find duplicates.

=head1 BUGS

Please report any bugs or feature requests to C<bug-file-same at rt.cpan.org>,  or
through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=File-Same>.
I will  be notified and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc File::Same

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=File-Same>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/File-Same>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/File-Same>

=item * Search CPAN

L<http://search.cpan.org/dist/File-Same/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2001 MessageLabs Limited.

This is free  software, you may use it and distribute  it under the same terms as
Perl itself.

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of File::Same


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'File::Same',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/File/Same.pm',
    LICENSE            => 'perl',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES     => {
        'Test::More'   => '1.001014',
    },
    PREREQ_PM          => {
        'strict'       => 0,
        'Digest::MD5'  => 0,
        'File::Spec'   => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'File-Same-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'File::Same' => { file => 'lib/File/Same.pm', version => '0.10' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/File-Same.git',
                web  => 'https://github.com/manwar/File-Same',
            },
        }})
     : ()
    ),
);
